evm_arithmetization/src/curve_pairings.rs

```diff
@@ -150,366 +141,937 @@ where
+        (true, false, true),
+        (false, false, true),
+        (false, true, true),
+        (false, false, true),
+        (false, true, true),
+        (false, true, true),
+        (false, false, false),
+        (true, true, true),
+        (true, false, true),
+        (true, false, true),
+        (false, true, true),
+        (true, false, true),
+        (false, true, true),
+        (false, true, true),
+        (true, true, false),
+        (true, true, false),
+        (true, true, false),
+        (true, false, false),
+        (false, false, true),
+        (true, false, false),
+        (false, false, true),
+        (true, false, true),
+        (true, true, false),
+        (true, true, true),
+        (false, true, true),
+        (false, true, false),
+        (true, true, true),
+    ];
+
+    const BN_EXPS2: [(bool, bool); 62] = [
+        (true, false),
+        (true, true),
+        (false, false),
+        (true, false),
+        (true, false),
+        (true, true),
+        (true, false),
+        (true, true),
+        (true, false),
+        (false, true),
+        (false, true),
+        (true, true),
+        (true, true),
+        (false, false),
+        (true, true),
+        (false, false),
+        (false, false),
+        (false, true),
+        (false, true),
+        (true, true),
+        (true, true),
+        (true, true),
+        (false, true),
+        (true, true),
+        (false, false),
+        (true, true),
+        (true, false),
+        (true, true),
+        (false, false),
+        (true, true),
+        (true, true),
+        (true, false),
+        (false, false),
+        (false, true),
+        (false, false),
+        (true, true),
+        (false, true),
+        (false, false),
+        (true, false),
+        (false, true),
+        (false, true),
+        (true, false),
+        (false, true),
+        (false, false),
+        (false, false),
+        (false, false),
+        (false, true),
+        (true, false),
+        (true, true),
+        (false, true),
+        (true, true),
+        (true, false),
+        (false, true),
+        (false, false),
+        (true, false),
+        (false, true),
+        (true, false),
+        (true, true),
+        (true, false),
+        (true, true),
+        (false, true),
+        (true, true),
+    ];
+
+    const BN_EXPS0: [bool; 65] = [
+        false, false, true, false, false, true, true, false, true, false, true, true, true, false,
+        true, false, false, false, true, false, false, true, false, true, false, true, true, false,
+        false, false, false, false, true, false, true, false, true, true, true, false, false, true,
+        true, true, true, false, true, false, true, true, false, false, true, false, false, false,
+        true, true, true, true, false, false, true, true, false,
+    ];
+}
+
+// The optimal Ate pairing implementation for BLS12-381 has been taken from
+// <https://github.com/zkcrypto/bls12_381>.
```

Comment:
Note that the library is under MIT/Apache2, but only includes default templates for both, so I couldn't include the MIT license mention here.